### DISCUSSION TITLE
BT816 ESP 8266 d1 mini

### DISCUSSION THREAD

**Discussion by homer2022** [RudolphRiedel/FT800-FT813#45] [Q&A]
Hello. I'm starting to play with Bt816. Could you please tell me how to connect bt816 to esp 8266, maybe some schematic and if you need to change something in your library files under ESP8266. How could you ask for some simple software for esp 8266 to check if it works properly. How to add a new display type to your config

**Comment by RudolphRiedel** [comment]
Currently I "only" have ESP8266 support for the Arduino framework implemented.
So check my PlatformIO/Arduino example:
https://github.com/RudolphRiedel/FT800-FT813/tree/5.x/examples/EVE_Test_Arduino_PlatformIO

I builds for a whole bunch of different Arduino targets, including a D1-Mini.
The configuration is done thru the platformio.ini file.

[env]
lib_deps = https://github.com/RudolphRiedel/FT800-FT813/archive/refs/heads/5.x.zip
framework = arduino
build_flags = -D EVE_EVE3_50G
check_tool = cppcheck
check_flags =
  cppcheck: --addon=cert.py --suppress=unusedFunction

[env:ESP8266_d1_mini]
platform = espressif8266
board = d1_mini
monitor_speed = 115200
build_flags =
        ${env.build_flags}
         -D EVE_CS=D2
         -D EVE_PDN=D1

So this currently builds for an EVE3-50G, uses the standard pins for SPI (whatever these are on the D1-mini),
plus D2 for Chip-Select and D1 for PowerDown.

According to the image I just found the SPI connects as:
D5 = SCK
D6 = MISO
D7 = MOSI

What a schematic for this setup looks like depends on what BT816 display you actually have,
An EVE3-35A (EVE_EVE3_35) for example has vastly different power requirements than a EVE3-70A (EVE_EVE3_70),

You probably can't use the 3V3 regulator on the ESP8266 board to power the display as well.
At least logic-level is not an issue, put in some series resistors like 47R and that is it.

**Comment by homer2022** [comment]
hello. could you give some simple example for testing under arduino ide. I want to connect such a display to bt 816 https://pl.aliexpress.com/item/32780373416.html?spm=a2g0o.cart.0.0.68de3c00eqSBTw&mp=1

**Comment by homer2022** [comment]
Will such an example be appropriate for the test?
#define EVE_CS  D8
#define EVE_PDN D1
#define BLPWM D3
//#define MOSI D7
//#define MISO D6
//#define SCK D5

#include <EVE.h>
#include <EVE_commands.h>
#include <EVE_config.h>
#include <EVE_cpp_wrapper.h>
#include <EVE_target.h>
#include <SPI.h>

uint8_t rez = 255;

void setup() {
  // put your setup code here, to run once:
  SPI.begin();
SPI.setFrequency(16000000); 
 Serial.begin(115200);
 pinMode(EVE_CS,OUTPUT);
 pinMode(EVE_PDN,OUTPUT);
 pinMode(BLPWM ,OUTPUT);

 rez = EVE_init();

 analogWrite(BLPWM ,1000);
 digitalWrite(EVE_PDN, LOW);

 EVE_cmd_dl(CMD_DLSTART); 
 EVE_cmd_dl(CLEAR_COLOR_RGB(0, 0, 255));
 EVE_cmd_dl(CLEAR(1,1,1));
 EVE_cmd_dl(DL_END);
 EVE_cmd_dl(DL_DISPLAY);
 EVE_cmd_dl(CMD_SWAP);
 EVE_execute_cmd();
}

void loop() {
  // put your main code here, to run repeatedly:
  delay(1000);
  Serial.print("Eve init: ");
  Serial.println(rez);
}

**Reply by RudolphRiedel** [reply]
> Will such an example be appropriate for the test?
> #define EVE_CS D8
> #define EVE_PDN D1

This will not work since EVE_target.h needs these.
But you can edit the default settings in EVE_target.h directly, see line 1810.

>I want to connect such a display to bt 816 https://pl.aliexpress.com/item/32780373416.html?spm=a2g0o.cart.0.0.68de3c00eqSBTw&mp=1

I have never tried something like this.

For this you need a configuration in EVE_config.h.
Like this:
----
#define EVE_TK035F5581_RTP

#if defined (EVE_TK035F5581_RTP)
#define EVE_HSIZE   (480L)  /* Thd Length of visible part of line (in PCLKs) - display width */
#define EVE_VSIZE   (320L)  /* Tvd Number of visible lines (in lines) - display height */

#define EVE_VSYNC0  (0L)    /* Tvf Vertical Front Porch */
#define EVE_VSYNC1  (10L)   /* Tvf + Tvp Vertical Front Porch plus Vsync Pulse width */
#define EVE_VOFFSET (22L)   /* Tvf + Tvp + Tvb Number of non-visible lines (in lines) */
#define EVE_VCYCLE  (360L)  /* Tv Total number of lines (visible and non-visible) (in lines) */
#define EVE_HSYNC0  (0L)     /* (40L)   // Thf Horizontal Front Porch */
#define EVE_HSYNC1  (41L)   /* Thf + Thp Horizontal Front Porch plus Hsync Pulse width */
#define EVE_HOFFSET (43L)   /* Thf + Thp + Thb Length of non-visible part of line (in PCLK cycles) */
#define EVE_HCYCLE  (548L)  /* Th Total length of line (visible and non-visible) (in PCLKs) */

#define EVE_PCLK    (6L)
#define EVE_PCLKPOL (1L)
#define EVE_SWIZZLE (0L)
#define EVE_CSPREAD (1L)
#define EVE_HAS_CRYSTAL
#define EVE_GEN 3

#EVE_TOUCH_RZTHRESH (1200L)
#endif
---

But this is where things get ugly.
There is no datasheet for this panel or even a type or what display controller it has, so all the timings are just a guess.
EVE_VSYNC1 , EVE_VOFFSET and EVE_VCYCLE  just maybe work as I set them.

EVE_PCLKPOL - this is a parameter of the panel.
EVE_SWIZZLE - this depends on how you connect the panel to the BT816 but the 24 data lines are
labeled D0 to D23 and therefore there is no indication what is R G and B.
EVE_HAS_CRYSTAL - comment out if you do not plan to use an external crystal
EVE_TOUCH_RZTHRESH -  try to increase this if the touch-panel seems to need too much force

On top I am looking at the provided schematic and I am not sure if this even can work.
At first it looks like a standard RGB interface.
But then it also has pins CS, RS, WR and RD which point to a parallel interface.

**Comment by homer2022** [comment]
Thanks for the help I have costakiego for this display. Will this example of what I gave you apart from configuring the cs pin will work. I need some post example on arduino I am going to check communication esp8266 with bt 816. What the config file should look like with the selected display EVE_TK035F5581_RTP
[R61529(3.5寸屏_芯片手册) (1).pdf](https://github.com/RudolphRiedel/FT800-FT813/files/8871494/R61529.3.5._.1.pdf)

**Reply by RudolphRiedel** [reply]
>Will this example of what I gave you apart from configuring the cs pin will work

Actually, no, but almost.

SPI.begin();
SPI.setFrequency(10000000); // this needs to be below 11 MHz for init
Serial.begin(115200);

pinMode(EVE_CS,OUTPUT);
pinMode(EVE_PDN,OUTPUT);
pinMode(BLPWM ,OUTPUT);
digitalWrite(EVE_PDN, LOW); // this puts EVE in power-down

rez = EVE_init();


That datasheet though is strange.
Most of the very few timings are set to "TBD".
I am not even sure if the chip kann work as a TFT with parallel RGB interface.
It kind of does say so but I am not sure, most of the information is for MIPI-DSI.
The internet has examples for a 8-bit parallel interface and I read that the chip
has it's own buffer-memory.

**Comment by homer2022** [comment]
Ok.Thanks for the help

**Comment by homer2022** [comment]
Hi. Can this display work with your library https://www.buydisplay.com/4-6-inch-color-bar-tft-lcd-iot-display-800x320-pixels-with-optl-touch-screen. If so, how should the file look like EVE_ config.h

**Reply by RudolphRiedel** [reply]
Generally speaking, yes, this should be possible.
Unfortunately there are no timing parameters in the so-called datasheet.

However, since these bar-type panels are just cut from regular panels, I would first try to use it as a common
800x480 display:

#if defined (EVE_ER-TFT046-1)
#define Resolution_800x480

#define EVE_PCLK    (2L)
#define EVE_PCLKPOL (1L)
#define EVE_SWIZZLE (0L)
#define EVE_CSPREAD (0L)
#define EVE_HAS_CRYSTAL
#define EVE_GEN 3
#define EVE_TOUCH_RZTHRESH  (1200L)
#endif

And again, EVE_PCLKPOL might be actually needed to be set to 0L, EVE_SWIZZLE  and EVE_HAS_CRYSTAL depend
on the actual design of your BT816 board.

**Comment by homer2022** [comment]
thank you

**Comment by homer2022** [comment]
Hello. Can this software work on this display https://en.aliexpress.com/item/32780373416.html?spm=a2g0o.cart.0.0.68de3c00eqSBTw&mp=1 with your libraries for bt816
#include <EVE.h>
#include <EVE_commands.h>
#include <EVE_config.h>
#include <EVE_cpp_wrapper.h>
#include <EVE_target.h>
#include <SPI.h>
#include "LCD.h"

#define EVE_CS  D8
#define EVE_PDN D1
#define BLPWM D2
//#define MOSI D7
//#define MISO D6
//#define SCK D5
#define SSPICS D3
#define SSPIMOSI D7
#define SSPISCK D5
#define LCDRST D4


uint8_t rez = 255;

void setup() {
  // put your setup code here, to run once:
pinMode(SSPICS,OUTPUT);
pinMode(SSPIMOSI,OUTPUT);
pinMode(SSPISCK,OUTPUT);
pinMode(LCDRST,OUTPUT);
Lcd_Initialize();
SPI.begin();
//SPI.setFrequency(1000000); 
 Serial.begin(115200);
 pinMode(EVE_CS,OUTPUT);
 pinMode(EVE_PDN,OUTPUT);
 pinMode(BLPWM ,OUTPUT);
// digitalWrite(EVE_CS, HIGH);
 //digitalWrite(EVE_PDN, HIGH);
// pinMode(D1,OUTPUT);
//digitalWrite(D1, LOW);
//delay(250);
// digitalWrite(D1, HIGH);
 rez = EVE_init();

 analogWrite(BLPWM ,1000);
 

 EVE_cmd_dl(CMD_DLSTART); 
 EVE_cmd_dl(CLEAR_COLOR_RGB(255, 0, 0));
 EVE_cmd_dl(CLEAR(1,1,1));
 EVE_cmd_dl(DL_END);
 EVE_cmd_dl(DL_DISPLAY);
 EVE_cmd_dl(CMD_SWAP);
 EVE_execute_cmd();

}
void loop() {

  // put your main code here, to run repeatedly:
  delay(1000);
  Serial.print("Eve init: ");
  Serial.println(rez);
}

LCD.cpp

#include "LCD.h"



void LCD_delay(uint16_t time)
{
  delay(time);
}

static void LCD_Reset(void)
{
  digitalWrite(LCDRST, LOW);
  LCD_delay(200);
  digitalWrite(LCDRST, HIGH);
  LCD_delay(200);
}

void LCD_WriteByteSPI(unsigned char dat)
{
  unsigned char n;

  for (n = 0; n < 8; n++)
  {
    if (dat & 0x80) {
      digitalWrite(SSPIMOSI, HIGH);
    }
    else {
      digitalWrite(SSPIMOSI, LOW);
    }
    dat <<= 1;
    digitalWrite(SSPISCK, LOW);
    digitalWrite(SSPISCK, HIGH);
  }
}
void SPI_WriteComm(uint16_t CMD)//9bit
{
  digitalWrite(SSPICS, LOW);
  digitalWrite(SSPIMOSI, LOW);
  digitalWrite(SSPISCK, LOW);
  digitalWrite(SSPISCK, HIGH);
  LCD_WriteByteSPI(CMD);
  digitalWrite(SSPICS, HIGH);
}

void SPI_WriteData(uint16_t tem_data)
{
  digitalWrite(SSPICS, LOW);
  digitalWrite(SSPIMOSI, HIGH);
  digitalWrite(SSPISCK, LOW);
  digitalWrite(SSPISCK, HIGH);
  LCD_WriteByteSPI(tem_data);
  digitalWrite(SSPICS, HIGH);
}
void Lcd_Initialize(void)	//LCD��ʼ������
{
  digitalWrite(SSPICS, HIGH);
  LCD_delay(20);
  digitalWrite(SSPICS, LOW);
  LCD_Reset();


  SPI_WriteComm(0xB4);
  SPI_WriteData(0x00);

  SPI_WriteComm(0xB0);
  SPI_WriteData(0x04);

  SPI_WriteComm(0x20);

  SPI_WriteComm(0x36);
  SPI_WriteData(0x6A);
  SPI_WriteComm(0x3A);
  SPI_WriteData(0x77);

  SPI_WriteComm(0xB3);
  SPI_WriteData(0x02);
  SPI_WriteData(0x00);
  SPI_WriteData(0x00);
  SPI_WriteData(0x20);



  SPI_WriteComm(0xc0);
  SPI_WriteData(0x03);
  SPI_WriteData(0xdf);
  SPI_WriteData(0x40);
  SPI_WriteData(0x12);
  SPI_WriteData(0x00);
  SPI_WriteData(0x01);
  SPI_WriteData(0x00);
  SPI_WriteData(0x55);

  SPI_WriteComm(0xc1);
  SPI_WriteData(0x07);
  SPI_WriteData(0x28);
  SPI_WriteData(0x08);
  SPI_WriteData(0x08);
  SPI_WriteData(0x00);

  SPI_WriteComm(0xc4);
  SPI_WriteData(0x70);
  SPI_WriteData(0x00);
  SPI_WriteData(0x03);
  SPI_WriteData(0x01);

  SPI_WriteComm(0xc6);
  SPI_WriteData(0x1d);

  SPI_WriteComm(0xc8);
  SPI_WriteData(0x06);
  SPI_WriteData(0x0c);
  SPI_WriteData(0x16);
  SPI_WriteData(0x24);
  SPI_WriteData(0x30);
  SPI_WriteData(0x48);
  SPI_WriteData(0x3d);
  SPI_WriteData(0x28);
  SPI_WriteData(0x20);
  SPI_WriteData(0x14);
  SPI_WriteData(0x0c);
  SPI_WriteData(0x04);
  SPI_WriteData(0x06);
  SPI_WriteData(0x0C);
  SPI_WriteData(0x16);
  SPI_WriteData(0x24);
  SPI_WriteData(0x30);
  SPI_WriteData(0x48);
  SPI_WriteData(0x3D);
  SPI_WriteData(0x28);
  SPI_WriteData(0x20);
  SPI_WriteData(0x14);
  SPI_WriteData(0x0C);
  SPI_WriteData(0x04);

  SPI_WriteComm(0xC9);
  SPI_WriteData(0x06);
  SPI_WriteData(0x0C);
  SPI_WriteData(0x16);
  SPI_WriteData(0x24);
  SPI_WriteData(0x30);
  SPI_WriteData(0x48);
  SPI_WriteData(0x3D);
  SPI_WriteData(0x28);
  SPI_WriteData(0x20);
  SPI_WriteData(0x14);
  SPI_WriteData(0x0C);
  SPI_WriteData(0x04);
  SPI_WriteData(0x06);
  SPI_WriteData(0x0C);
  SPI_WriteData(0x16);
  SPI_WriteData(0x24);
  SPI_WriteData(0x30);
  SPI_WriteData(0x48);
  SPI_WriteData(0x3D);
  SPI_WriteData(0x28);
  SPI_WriteData(0x20);
  SPI_WriteData(0x14);
  SPI_WriteData(0x0C);
  SPI_WriteData(0x04);

  SPI_WriteComm(0xCA);
  SPI_WriteData(0x06);
  SPI_WriteData(0x0C);
  SPI_WriteData(0x16);
  SPI_WriteData(0x24);
  SPI_WriteData(0x30);
  SPI_WriteData(0x48);
  SPI_WriteData(0x3D);
  SPI_WriteData(0x28);
  SPI_WriteData(0x20);
  SPI_WriteData(0x14);
  SPI_WriteData(0x0C);
  SPI_WriteData(0x04);
  SPI_WriteData(0x06);
  SPI_WriteData(0x0C);
  SPI_WriteData(0x16);
  SPI_WriteData(0x24);
  SPI_WriteData(0x30);
  SPI_WriteData(0x48);
  SPI_WriteData(0x3D);
  SPI_WriteData(0x28);
  SPI_WriteData(0x20);
  SPI_WriteData(0x14);
  SPI_WriteData(0x0C);
  SPI_WriteData(0x04);

  SPI_WriteComm(0xD0);
  SPI_WriteData(0x95);
  SPI_WriteData(0x0A);
  SPI_WriteData(0x08);
  SPI_WriteData(0x10);
  SPI_WriteData(0x39);

  SPI_WriteComm(0xD1);
  SPI_WriteData(0x02);
  SPI_WriteData(0x2c);
  SPI_WriteData(0x2c);
  SPI_WriteData(0x44);
  SPI_WriteData(0x00);

  SPI_WriteComm(0x11);
  LCD_delay(100);
  SPI_WriteComm(0x29);
  LCD_delay(100);
  SPI_WriteComm(0x2C);

  SPI_WriteComm(0x36);
#if LCD_RGB_ORIENTATION
  SPI_WriteData(0x09);
#else
  SPI_WriteData(0x08);
#endif
}


LCD.h

#include"Arduino.h"
#define SSPICS D3
#define SSPIMOSI D7
#define SSPISCK D5
#define LCDRST D4

#define LCD_RGB_ORIENTATION      1  




void Lcd_Initialize(void); 
void LCD_delay(uint16_t time);
void SPI_WriteComm(uint16_t CMD);
void SPI_WriteData(uint16_t dat);
void LCD_WR_REG(uint16_t Index,uint16_t CongfigTemp);

**Reply by RudolphRiedel** [reply]
I am not really sure what your code is supposed to be doing, I dusted off my D1 mini and tried it but it looks strange in the logic analyzer.
Please explain what you are trying to do.

I also did however tryout my Arduino demo with the D1 mini and an EVE3-50G and found it to be not working.
There was a bug in EVE_cpp_wrapper.cpp from the time I implemented it almost a year ago, 
__builtin_bswap32() was used in wrapper_spi_transmit_32() a second time with the first
time in EVE_target.h / spi_transmit_32() for the ESP8266.

I fixed that, uploaded the file, changed my platformio.ino to use D3 for CS and D4 for PDN and now it works.

That however brings us back to the timing data.
There is no specifcation for the display whatsoever.
I have no display with 480x320 in my EVE_config.h.
These might be cut from 800x480 panels.

Based on some other completely unrelated display I just found with these dimensions,
this might work:
#if defined (EVE_35_320x480)
#define EVE_HSIZE   (320L)
#define EVE_VSIZE   (480L)
#define EVE_VSYNC0  (0L)
#define EVE_VSYNC1  (4L)
#define EVE_VOFFSET (6L)
#define EVE_VCYCLE  (486L)
#define EVE_HSYNC0  (0L)
#define EVE_HSYNC1  (10L)
#define EVE_HOFFSET (15L)
#define EVE_HCYCLE  (335L)
#define EVE_PCLK    (7L)
#define EVE_PCLKPOL (1L)
#define EVE_SWIZZLE (0L)
#define EVE_CSPREAD (0L)
//#define EVE_HAS_CRYSTAL
#define EVE_GEN 3
#endif

EVE_PCLKPOL with a 50:50 chance.
And EVE_SWIZZLE need to be adjusted to how the display is actually wired to the BT816.

**Comment by homer2022** [comment]
This is a test program I want to display the background in one color to check if bt is controlling this display

**Reply by RudolphRrr** [reply]
Ok, but what is Lcd_Initialize(); about?
Is this for another completely unrelated display?

**Comment by homer1966** [comment]
it is for this display.this display works properly with the TK499 board

**Reply by RudolphRiedel** [reply]
I had no idea what a TK499 board is.
So I guess the extra SPI code is for the display-controller on the TFT?
It has to be configured?
This makes things even more complicated.

But if there is a software for this display that runs with this board, what timing parameters are used?

**Comment by homer2022** [comment]
this display works correctly on this firmware https://megawrzuta.pl/download/6a44b6b322582d185832f3e7fd40bca4.html.
tak dodatkowy kod dotyczy kontrolera wyswietlacza i trzeba go skonfigurowac. tak jest zrobione w tym oprogramowaniu

**Comment by homer1966** [comment]
this display works correctly on this firmware https://megawrzuta.pl/download/6a44b6b322582d185832f3e7fd40bca4.html.
this extra code applies to the display controller and must be configured. so it is done in this software

**Reply by RudolphRiedel** [reply]
I see, so in the case of using the TFT with the ESP8266 and a BT816, the display-controller on the TFT would be just annother SPI peripheral.
That should be working with the SPI class as well, it may need different initialization though if
 SPI.beginTransaction(SPISettings(8000000, MSBFIRST, SPI_MODE0));
does not work.

Like this:
SPI.beginTransaction(SPISettings(2000000, LSBFIRST, SPI_MODE2));
init_TK035F5589()
SPI.endTransaction()
SPI.beginTransaction(SPISettings(8000000, MSBFIRST, SPI_MODE0));
TFT_init();

Do you have the TKM32F499 board?
I would use a logic-analyzer to trace the SPI traffic and then try to just mimic it.

I found the Arduino core for the TKM32F499 yesterday which includes the examples.

**Comment by homer1966** [comment]
It seems to me that for proper work it is also necessary to control after sleep with esp 8266 This is probably done on this tk499 module. I noticed during the measurements with the oscilloscope that the pclk mileage has a frequency of 21mhz, vsync 113hz and hsync 57.1khz, these are values ​​about 2 times larger than on the same pins of the display connected to the bt816 board. The order of the rgb signals is consistent

**Comment by RudolphRiedel** [comment]
I had another look.
The R61529 seems to have at least two compatible replacements, ILI9488 and HX8357.
The interface is actually called "DBI Type C" and there is a 3wire version and a 4wire version.
The 3wire version is using CSX, WRX_SCL and SDA with 9 bits of data and the first bit is ZERO for a command and ONE for data.
The 4wire version is using DCX additionally and 8 bits of data, with DCX set to ZERO for a command and to ONE for data.
SDA is bi-directional.
So this looks a lot like SPI but with extra steps.

From the timing informations the R61529 should have a maximum WRX_SCL clock of 10MHz while the ILI9488 and HX8357 go up to 15MHz.

Interesting but it's not like I am really getting anywhere.

SPI_WriteComm(0x3A); 
SPI_WriteData(0x77);

This is setting the DPI and DBI pixel formats to 24 bit.

And MIPI DPI appears to be the normal TFT parallel pixel interface that the EVE chips are using.
I have no idea though what the minimum required sequence would be to configure the chip and use it in MIPI DPI mode.

However, assuming that the sequence in your code is correct, the ported version should generate
the apropriate bit patterns. (the example from the TK499 Arduino library has a different order of commands).

SPI_WriteComm(0xB4); -> 0 1011 0100
SPI_WriteData(0x00); -> 1 0000 0000
SPI_WriteComm(0xB0); -> 0 1011 0000
SPI_WriteData(0x04); -> 1 0000 0100

**Comment by homer1966** [comment]
So it can't be done

**Reply by RudolphRiedel** [reply]
> So it can't be done

What makes you say that?
The TKM32F499 or TK499  board is doing it already, only using an internal graphics controller.
From the TKM32F499 datasheet:

⚫ LCD-TFT controller, support 16 million colors
RGB888, support double page switching, out-of-order 4K
address alignment

The LCD-TFT display controller provides 24-bit parallel digital RGB (red, green, blue), and transmits all signals
interface directly with a wide range of 16 million color RGB888 LCD and TFT panels, it has the following features:
⚫ Support standard hsync/vsync format digital video output
⚫ Digital video output timing adjustable
⚫ The output format supports up to RGB888
⚫ Internal 2 x 480x64 dpram

This controller does not have a MIPI interface either, the integrated display controller is using the same signals as the EVE chip does.

I like to know more though, I am not comfortable with just copying a sequence like this:
SPI_WriteComm(0xB4);
SPI_WriteData(0x00);
SPI_WriteComm(0xB0);
SPI_WriteData(0x04);
SPI_WriteComm(0x20);
SPI_WriteComm(0x36);
SPI_WriteData(0x6A);
SPI_WriteComm(0x3A);
SPI_WriteData(0x77);

Without knowing what it does.

And without knowing what the signals are supposed to look like there is no telling if the code is correct.

**Comment by homer1966** [comment]
Will any config in your library be able to work with this display https://pl.aliexpress.com/item/32639735669.html.This 3.5 inch display I was able to run n from your libraries only in the main code you had to initialize bt816 earlier and then the display itself thanks for your help

**Reply by RudolphRiedel** [reply]
This looks plausible in regards of the signals.
But then again, there is no documentation for this display and I have no information in regards of your BT816 board.
If your BT816 has a crystal populated then changing the demo to the EVE_EVE3_50 configuration might work.

I have a ESP8266 up and running here with an EVE_EVE3_50G which is using almost the same configuration but with a BT815 and a touch-controller that needs a few lines extra.

**Comment by homer2022** [comment]
this display works on this driver https://community.nxp.com/pwmxy87654/attachments/pwmxy87654/imx-processors/115314/1/LG4573A_DS_V1.1.5.pdf

**Reply by RudolphRrr** [reply]
This is interesting, 1750 pins, wth. :-)
This controller "only" has MIPI DPI for pixel data and a SPI or MIPI DBI interface for register access.
It might just work in the default configuration using only the MIPI DPI aka RGB888 interface.

**Comment by homer2022** [comment]
does it mean that it is enough to connect its inputs rgb and vsync, hsync, de, pclk with analogous outputs from bt 816 the display should work if it has the correct config file

**Reply by RudolphRrr** [reply]
Apart from that the RGB outputs are digital and not analog, I am not exactly sure that this works but I am pretty confident that this might work.
The VM816C59A-N module from Bridgetek should be a good reference to check if all necessary signals are present at the connector of this display: https://brtchip.com/ic-module/wp-content/uploads/sites/3/2022/01/DS_VM816C.pdf

The manual for the EVE3 modules from Matrix Orbital also have a schematic.


I chose to not build my own EVE board and try to connect some TFT to it.
When checking out TFTs I quickly learned that there is no standard connector.
Not only do these use different numbers of pins to do the same thing, the pinout varies and so does the length of the cables.

Then there is that you get no documentation for any of the cheaper TFTs.
And even if you do get a datasheet, most of the time important parameters are missing.
I am in no position to order TFTs to my specification so I would end up with reworking my EVE board over and over again.

So yes, EVE modules from Matrix Orbital, Riverdi, Crystalfontz, Newhaven or 4D-Systems  are more expensive than the naked grey-market TFTs one could get.
But these are designed to work with EVE, come with documentation and support.

**Comment by homer1966** [comment]
Thank you for the information, I will try to run the second display, if I can, I will let you know. How to display a logo on the display. You could enter some simple code for esp826 or point to a page where such examples can be found

**Reply by RudolphRiedel** [reply]
My PlatformIO Arduino example is building for ESP8266 as well:
https://github.com/RudolphRiedel/FT800-FT813/tree/5.x/examples/EVE_Test_Arduino_PlatformIO

The platformio.ini file has the settings for the various boards.

The general part for all boards:
[env]
lib_deps = https://github.com/RudolphRiedel/FT800-FT813/archive/refs/heads/5.x.zip
framework = arduino
build_flags = -D EVE_EVE3_50G
check_tool = cppcheck
check_flags =
  cppcheck: --addon=cert.py --suppress=unusedFunction

And specific to ESP8266:
[env:ESP8266_d1_mini]
platform = espressif8266
board = d1_mini
monitor_speed = 115200
build_flags =
        ${env.build_flags}
         -D EVE_CS=D3
         -D EVE_PDN=D4

When you just hit "build" on the button the example is compiled for all boards:
Environment                Status    Duration
-------------------------  --------  ------------
uno                        SUCCESS   00:00:01.846
avr_pro                    SUCCESS   00:00:02.165
nano328                    SUCCESS   00:00:02.036
mega2560                   SUCCESS   00:00:02.061
adafruit_metro_m4          SUCCESS   00:00:22.934
samd21_m0-mini             SUCCESS   00:00:13.682
ESP32                      SUCCESS   00:02:23.050
ESP8266_d1_mini            SUCCESS   00:00:02.292
nucleo_f446re              SUCCESS   00:00:10.521
teensy41                   SUCCESS   00:00:45.408
adafruit_feather_nrf52840  SUCCESS   00:00:06.644
bbcmicrobit_v2             SUCCESS   00:00:03.050
teensy35                   SUCCESS   00:00:03.992
xmc1100_xmc2go             SUCCESS   00:00:05.066

**Comment by homer1966** [comment]
Thank you, Mie rather, it was about a code that would be part of another code because I wanted to do in this display that I started to do so that after running for some time the logo would be displayed. Instead of a uniform background, there was a logo

**Reply by RudolphRrr** [reply]
Well now, isn't that an issue for another day? :-)
When the demo is up and running you can make it display whatever you like.

**Reply by homer2022** [reply]
Is it possible to run on your library and on the BT816 a display with a resolution of 960x320

**Reply by RudolphRiedel** [reply]
There is nothing in my library that I am aware of that would prevent this.
Finding the correct timing parameters might be tricky though since it is rather likely that a display with this resolution is cut from a larger panel so odd offsets might be required as well as using VSIZE and HSIZE values other than the useable resolution.
Matrix-Orbital has a couple of bar type displays that are cut from larger panels, the EVE3x-39G would be one with 480x128.
And calibrating such a display might be not as straight forward but this is where EVE_calibrate_manual() comes in handy.

**Reply by homer2022** [reply]
Thank you for the information

**Reply by homer1966** [reply]
Where can I find this EVE_calibrate_manual() and how do I use it? How can I invert the colors

**Reply by RudolphRiedel** [reply]
It is at the end of EVE_commands.c:
void EVE_calibrate_manual(uint16_t width, uint16_t height);

And the colors are fixed.

**Comment by homer1966** [comment]
I have such a display https://www.alibaba.com/product-detail/4-inch-TFT-Screen-3-99_1600444067823.html?spm=a2700.shop_plgr.89.12.4b6832deYoSkAC . I managed to work it on the 480x800 configuration But nir reproduces colors correctly. Just like in the picture. Works on the RGB 16 bit interface. What could be the cause of the wrong work. I am attaching my code and display data. How would you see it
![IMG_20230211_085554](https://user-images.githubusercontent.com/59050826/218320675-702c72a5-e49a-45f9-8f6d-8fd6a0fe6554.jpg)
[H40G16-00Z Ver00  Spec.pdf](https://github.com/RudolphRiedel/FT800-FT813/files/10716416/H40G16-00Z.Ver00.Spec.pdf)

**Comment by homer1966** [comment]
[wyswietlacz-3.99_cala.zip](https://github.com/RudolphRiedel/FT800-FT813/files/10716429/wyswietlacz-3.99_cala.zip)

**Comment by RudolphRiedel** [comment]
This is interesting.
Unfortunately there are no timing informations in the "datasheet".
Checking the datasheet of the ST7701S display controller that is used on the panel shows that it supports a resolution of 480 x 960 among others, so this panel is probably cut down to 400 x 960 from such a panel.
Looks like these are smartphone panels.
Anyways, most likely the configuration for 800x480 has a line that is too short.

I see the function Lcd_Initialize() in your example package but decoding it to find out what parameters you use is something I rather not spend my evening with. :-)

What are your values for HSIZE, HCYCLE, HOFFSET, HSYNC0 and HSYNC1?

I would try something like this:  
HSIZE = 960  
HCYLCE = 1024  
HSYNC0 = 0  
HSYNC1 = 10  
HOFFSET = 70  

I looked for other displays using 960 x whatever to check for timing parameters but I found close to nothing.
Either there was no datasheet provided or the "datasheet" contained little to no information.

**Reply by homer2022** [reply]
OK thank you for the information
